many code improvements in ui/server as I walk through the server code

daattali opened the pull request:
The server file is huge and hard to take in at once, so I'm walking through it to try to understand the logic. As I'm going through, I'm refactoring anything I come across that can be made simpler/cleaner/more efficient

StevenLShafer commented:
Dear Dean:

Wonderful!

If you want to break it up into modules, that would be fine by me. I agree it is very unwieldly. Unfortunately, the I found the description of shiny modules quite opaque. However, I was aware that eventually server.R would need to be broken into smaller functional modules, so doing that would be hugely helpful.

Thanks,

Steve

From: Dean Attali <notifications@github.com>
Sent: Monday, March 2, 2020 4:54 PM
To: StevenLShafer/stanpumpR <stanpumpR@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: [StevenLShafer/stanpumpR] many code improvements in ui/server as I walk through the server code (#26)


The server file is huge and hard to take in at once, so I'm walking through it to try to understand the logic. As I'm going through, I'm refactoring anything I come across that can be made simpler/cleaner/more efficient

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/StevenLShafer/stanpumpR/pull/26

Commit Summary

  *   many code improvements in ui/server as I walk through the server code

File Changes

  *   A R/server-helpers.R<https://github.com/StevenLShafer/stanpumpR/pull/26/files#diff-f9c4ae43888ced8ab9a3ddc2a82082f2> (28)
  *   M R/validateTime.R<https://github.com/StevenLShafer/stanpumpR/pull/26/files#diff-c27a5fd1436ac10ab9f6ba25299afe58> (14)
  *   M global.R<https://github.com/StevenLShafer/stanpumpR/pull/26/files#diff-e69f1fd994914230b6ee68b5a5a24a9e> (15)
  *   M server.R<https://github.com/StevenLShafer/stanpumpR/pull/26/files#diff-b388ce12a8876fa2ab1c7c71a5ca15e7> (293)
  *   M ui.R<https://github.com/StevenLShafer/stanpumpR/pull/26/files#diff-345025a688475250d13a1e0f6050f07e> (108)
  *   M www/app.css<https://github.com/StevenLShafer/stanpumpR/pull/26/files#diff-0fdbe601bbf5f40f90f60f3e33ce45f2> (23)
  *   A www/app.js<https://github.com/StevenLShafer/stanpumpR/pull/26/files#diff-4e0a58cc8477966840e5de1fdd7df1ea> (14)
  *   D www/cancel..jpg<https://github.com/StevenLShafer/stanpumpR/pull/26/files#diff-ca0477a1393c58d4703bbbd1e3909fda> (0)
  *   D www/shinyjs-funcs.js<https://github.com/StevenLShafer/stanpumpR/pull/26/files#diff-77b841d5222e8cb6464f569c127794ab> (4)

Patch Links:

  *   https://github.com/StevenLShafer/stanpumpR/pull/26.patch
  *   https://github.com/StevenLShafer/stanpumpR/pull/26.diff

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/StevenLShafer/stanpumpR/pull/26?email_source=notifications&email_token=AFHWJTYQCVDAE6RFA447473RFRILBA5CNFSM4K772U22YY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4IR32J2Q>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AFHWJT7YO52BNRRRO6U2HHDRFRILBANCNFSM4K772U2Q>.

daattali commented on R/server-helpers.R:
```diff
@@ -0,0 +1,28 @@
+    )
+  )
+}
+
```
Because I was overwhelmed by the size of the server file, I tried moving chunks of code that didn't have to belong in the server outside, like `showIntroModal()` and `getReferenceTime()`

daattali commented on global.R:
```diff
@@ -121,10 +121,23 @@ doseTableInit <- data.frame(
+      text <- argslist[[1]]
+    } else {
+      text <- paste(argslist, collapse = "")
+    }
```
I made `outputComments()` to work with any number of arguments, so instead of doing `outputComments(paste("hello", "there"))` you can just do `outputComments("hello", "there")`

daattali commented:
@StevenLShafer This PR is ready - but before merging, I suggest pulling the changes locally and testing them. I tested as best I could, but I don't know all the hidden features of this app so I'm sure there are some code paths I haven't tested. 

I've spent a few days going through the server code and adding a lot of reactivity. It seems like a lot of the code was written in a more imperative fashion and didn't really embrace reactivity.  What I mean by that is in reactive programming (as opposed to imperative programming), the idea is that you shouldn't have to use a lot of flags and comparing current to previous value in order to determine when to re-run something - if reactivity is done correctly, the beauty of it is that it takes care of it automatically. I've fixed many variables to be reactive, and introducing more reactivity also ended up fixing some bugs organically. Here are my comments explaining the reason behind a lot of my changes, it's A LOT to read through so take your time. I'll **bold** the few that I think are very important and can be used for learning. The last 2 commits (6 and 7) I think would be the most beneficial to understand.

### [Commit 1](https://github.com/StevenLShafer/stanpumpR/pull/26/commits/12a1b5122d0ee5b2c6c40872e251f620e45091a0) - general non-shiny-related cleanup

- Because I was overwhelmed by the size of the server file, I tried moving chunks of code that didn't have to belong in the server outside, like `showIntroModal()` and `getReferenceTime()`
- I made `outputComments()` to work with any number of arguments, so instead of doing `outputComments(paste("hello", "there"))` you can just do `outputComments("hello", "there")`
- I saw that you had to explicitly remove all the shinyjs delay events from bookmarking, I apologize for the delay function being so messy! I rewrote the code that auto scrolls the logger so that it no longer uses delay and your sessions doesn't get dirtied with all those inputs
- **You should almost never place a `render*()` function inside an observer or an isolate. There are a few places where I had to move some code around to fix that. Doing that results in logically inconsistent behaviour; for example [here](https://github.com/StevenLShafer/stanpumpR/pull/26/commits/12a1b5122d0ee5b2c6c40872e251f620e45091a0#diff-b388ce12a8876fa2ab1c7c71a5ca15e7L162-L166) by placing the render inside an isolate, did you want it to never run (and thus used an isolate)? That's not what happens**
- Wherever possible, when dealing with dynamic UI, I moved as much code into the UI and have code in the server for the part that truly must be calculated dynamically. For example the [code to create the reference time](https://github.com/StevenLShafer/stanpumpR/pull/26/commits/12a1b5122d0ee5b2c6c40872e251f620e45091a0#diff-b388ce12a8876fa2ab1c7c71a5ca15e7L221-L279) moved mostly into the UI, and in the server all I'm doing is updating the dropdown to the correct time
- Whenever I saw chunks of code used in multiple places, I created a variable for it instead ([example](https://github.com/StevenLShafer/stanpumpR/pull/26/commits/12a1b5122d0ee5b2c6c40872e251f620e45091a0?w=1#diff-b388ce12a8876fa2ab1c7c71a5ca15e7R258))
- Made some fixes to the UI like moving all inline CSS into a CSS file, rewrote the HTML+javascript hack that was used for telling the server what reference time to use

### [Commit 2](https://github.com/StevenLShafer/stanpumpR/pull/26/commits/8d45bbd8a3a0a200be97d77e4807296414297786) - starting to break down the main observe loop

- I saw there were some "magic numbers" that were used in multiple places (min/max age/weight/height). I chose to use variables instead of hardcoding the numbers, which eliminates the potential error of updating a number in some places but not others (which actually was happening previously - one of the checks for weight was checking for a minimum of 1, while the others were using 0.1) 
- Started moving logic out of the huge main observation look, like [having a dedicated function to check for covariates](https://github.com/StevenLShafer/stanpumpR/pull/26/commits/8d45bbd8a3a0a200be97d77e4807296414297786?w=1#diff-f9c4ae43888ced8ab9a3ddc2a82082f2R29-R52) 
- Got rid of `PK_set` reactiveVal
- The checks for covariates weren't taking into account the units of the age/weight/height
- Moved recalculating drug PKs out of the main observer loop and into [its own function](https://github.com/StevenLShafer/stanpumpR/pull/26/commits/8d45bbd8a3a0a200be97d77e4807296414297786?w=1#diff-f9c4ae43888ced8ab9a3ddc2a82082f2R106-R128) - but notice that I really didn't like that this was modifying the `drugs` input so I made an explicit comment saying that this function has that side effect so we can beware of that. In a later commit I address this by changing it from a `reactiveValues` to a `reactiveVal` 

### [Commit 3](https://github.com/StevenLShafer/stanpumpR/pull/26/commits/825c94491dcea156ea9587860fddba5a6d0b6837) - outputComments can look at a special environment option to determine if to echo or not, so that you don't have to pass `echo = DEBUG` into every single call

### [Commit 4](https://github.com/StevenLShafer/stanpumpR/pull/26/commits/16133e9e6930538a72b9c67828a045ee1749680b) - misc cleanup, removing unused variables/flag/code, move code away from server

### [Commit 5](https://github.com/StevenLShafer/stanpumpR/pull/26/commits/8f8ca74895fe17f7bcfda3267842ec9fd8092bac?w=1) - simplify email button code

### [Commit 6](https://github.com/StevenLShafer/stanpumpR/pull/26/commits/d68874298d3fc01bb53505c9d18cb3bdeee3f68c) - big reactivity cleanup

- **Removed all instances of  `<<-` and replace them with reactives. This is important because with `<<-` it's not clear when/where variables are modified, the variable breaks away from the reactivity graph so now anything that uses it doesn't automatically get updated, and instead you end up using flags and comparisons to try to figure out if the variable got updated. Another reason it's bad is that it can save the object in the global environment which is shared between all users, which means that one user's actions will overwrite another user**
- **I noticed that `drugDefaults` was a variable that gets modified and you always wanted to use its latest value, so that should scream "reactive variable". It was used as a non-reactive variable that gets overwritten and having a flag to keep track of whether it changed, and any code that depended on it didn't explicitly know when it changed. That's exactly what reactive programming solves - it takes care of all this naturally. For example, just below the definition for drugDefaults there was a variable `drugList <- drugDefaults$Drug` and it wasn't clear to anyone looking at the code whether this is meant to always update every time drugDefaults updates or if it's just a one-time thing (I had to look through the code to figure that out, and make `drugList` reactive as well)**
- Fixed a bug that was caused by `<<-` overwriting globally: when one person clicks "Edit Drugs" and changed the units, the new units got reflected in all current users sessions. Only when the app died, the units for everyone got reset. I didn't have to do anything special to fix the bug, when I changed `drugDefaults` to be a `reactiveVal()` and implemented it as a reactive everywhere, this got taken care of automatically
- Another bug got fixed by making `drugDefaults` a reactive is that previously every time after I edited drugs to change the units, the list of units in the dose table on the left side kept reverting back to the initial units. After making `drugDefaults` reactive, it naturally fixed that problem.
- I mentioned this before, but render functions should never be inside observers. The entire "Edit Drugs" table output was defined inside an observer, and that observer was doing too much. I changed it so that the observer only shows the modal, but the logic for how to create the table is independent

### [Commit 7](https://github.com/StevenLShafer/stanpumpR/pull/26/commits/a4418eaf2891eeeb0e23446664ee8a72d1819a73) - use more reactive programming, convert `drugs` to a `reactiveVal`

- Similar to how I had to convert drugDefaults to a regular reactiveVal, I had to do the same to `drugs` so that it would be much easier to logically follow what happens to that variable
- **To expand on the previous point: You should always use `reactiveVal` rather than `reactiveValues` as much as possible because reactiveValues allows you to modify it from anywhere which is bad practice. You can't do that with reactiveVal, so you're bound to use safer coding. I changed `drugs` to be a `reactiveVal` for this reason. When I was looking through the server code to see where `drugs` is being modified, I couldn't only look at where `drugs$... <-` happens because it was able to get modified from anywhere. Because it's a reactiveValues, I had to look everywhere that `drugs` was passed in as an argument and see if it gets modified there. That's why side effects of functions is a bad practice in reactive programming, because you're breaking away from the reactivity graph and now it's unclear where things get modified. There was no easy way to know where `drugs` was being modified which means it's very difficult and bug-prone to try to change any code that uses it. **


---

There still is work left to be done - I think the "main observation loop" can be broken down to be much smaller and there are many large chunks of code that are doing too much and should be doing one specific task and logic should be broken up so that every function has one clear goal. I also think the three reactiveVals `plotObjectReactive` `allResultsReactive` and `plotResultsReactive` should be able to get reworked so that they're not needed. I also think the entire `prior` reactive might not be needed, because it essentially tries to mimic the idea of reactivity but doing it manually instead of automatically. Everything I mentioned in this paragraph I've done already for a few cases, but there are more left. I think the last 2 commits are good examples of the direction the code should go, how to remove a lot of the complex code and make it reactive instead. I really think the entire `prior` would become obsolete. Shiny's reactivity is really an amazing feature, but it's also very unintuitive unfortunately!

daattali commented:
I wanted to try to be more clear in why I think leveraging reactivity correctly can great simplify the code and make it easier to maintain. I built the toy example below that tries to mimic on a small scale the general idea of the main observation loop:

```
library(shiny)
library(ggplot2)

makePlot <- function(xvar, yvar) {
  ggplot(iris, aes_string(xvar, yvar)) + geom_point()
}

ui <- fluidPage(
  selectInput("xvar", "X variable", choices = names(iris), selected = names(iris)[1]),
  selectInput("yvar", "Y variable", choices = names(iris), selected = names(iris)[2]),
  plotOutput("plot")
)

server <- function(input, output, session) {
  # This list holds all the previous values
  prior <- reactiveValues(
    xvar = "",
    yvar = ""
  )
  
  # This will contain the current plot
  plotObject <- reactiveVal()
  
  # Main workhorse - when any inputs change, determine how to update the plot
  observe({
    replotAxesFlag <- FALSE
    if (prior$xvar != input$xvar) {
      prior$xvar <- input$xvar
      replotAxesFlag <- TRUE
    }
    if (prior$yvar != input$yvar) {
      prior$yvar <- input$yvar
      replotAxesFlag <- TRUE
    }
    
    if (replotAxesFlag) {
      newPlot <- makePlot(prior$xvar, prior$yvar)
      plotObject(newPlot)
    }
  })
  
  output$plot <- renderPlot({
    plotObject()
  })
}

shinyApp(ui, server)
```

That can be simplified to

```
library(shiny)
library(ggplot2)

makePlot <- function(xvar, yvar) {
  ggplot(iris, aes_string(xvar, yvar)) + geom_point()
}

ui <- fluidPage(
  selectInput("xvar", "X variable", choices = names(iris), selected = names(iris)[1]),
  selectInput("yvar", "Y variable", choices = names(iris), selected = names(iris)[2]),
  plotOutput("plot")
)

server <- function(input, output, session) {
  plotObject <- reactive({
    makePlot(input$xvar, input$yvar)
  })
  
  output$plot <- renderPlot({
    plotObject()
  })
}

shinyApp(ui, server)
```

Of course your case is inherently more complex so this code may look silly, but it's the same general idea.

trshafer commented on server.R:
```diff
@@ -9,40 +9,14 @@
-      size="s"
-    )
-  )
+  showIntroModal()
```
This function has side effects in the UI. Should the direct UI code stay in server.R? I could see this being `showModal(introModal())`. YMMV

daattali replied:
It's a bit strange and unintuitive, but the current way is the right way. I know it looks strange, but modals are one of shiny's quirks

trshafer commented on server.R:
```diff
@@ -1605,7 +1301,7 @@ dblclickPopupDrug <- function(
         label = "Units",
         choices = c(units),
-        selected = drugDefaults$Bolus.Units[i],
+        #selected = drugDefaults()$Bolus.Units[i],
```
remove this entirely?

daattali replied:
Yes it can be removed. I'm happy to see you're going through my code!

There's actually other places in the code that also make references to `i` that's not defined in local scope. This code used to work because `i` was defined as a leftover from a random for loop earlier but that for loop no longer exists. The only reason I noticed this one is because I actually noticed an error, but there must be other places.

trshafer commented:
I pulled these changes down and ran some manual validation. It looks good to me. The travis builds were failures on travis's side. Restarted the builds and they passed.

trshafer commented:
Code looks good. Merging and deploying with upcoming automatic deployment PR.
